### ISSUE TITLE
RFC: Define what is output timezone

### ISSUE THREAD

**Issue by chinpei215** [cakephp/cakephp#8948]
This is a (multiple allowed):
* [ ] bug
* [ ] enhancement
* [x] feature-discussion (RFC)

* CakePHP Version: master
* Platform and Target: non-related

### What you did

I noticed that #8935 introduced several annoying issues into cake.

#### 1. Date class doesn't support default output timezone

If I execute the following code at Jun 7, 2016, 3:00 AM GMT+9 (Japan Standard Time):
```php
date_default_timezone_set('UTC');
Time::setDefaultLocale('ja_JP');

Time::setDefaultOutputTimezone('Asia/Tokyo');
echo new Time();
```
It would display `2016/06/07 3:00` as I expected.

But I cannot do similar works by using `Cake\I18n\Date` class.
```php
date_default_timezone_set('UTC');
Date::setDefaultLocale('ja_JP');

Date::setDefaultOutputTimezone('Asia/Tokyo'); 
echo new Date();
```

It would throw an exception because `Cake\I18n\Date` class doesn't support the `setDefaultOutputTimezone` method. And if I don't call the method, it will display `2016/06/06` not `2016/06/07`.

#### 2. The json_encode() function is effected by default output timezone

After I set default output timezone, calling `json_encode()` displays datetimes with the default output timezone.

```php
$time = new Time();
echo json_encode($time);
// "2016-06-06T18:00:00+0000"

Time::setDefaultOutputTimezone('Asia/Tokyo');

echo json_encode($time);
// "2016-06-07T03:00:00+0900"
```

This means that if my application stores entities into some storage with JSON, the representation of datetime fields are effected by user preferences.

#### 3. Logging methods are effected by default output timezone

Similar to the above. `Cake\Log\Log::write()` and other logging methods are effected by the default output timezone.

```php
$time = new Time();
Log::write('debug', $time);
// Debug: 2016/06/06 18:00

Time::setDefaultOutputTimezone('Asia/Tokyo');

Log::write('debug', $time);
// Debug: 2016/06/07 3:00
```

#### 4. FormHelper doens't use default output timezone

If I execute the following code at Jun 7, 2016, 3:00 AM GMT+9 (Japan Standard Time):
```php
Time::setDefaultOutputTimezone('Asia/Tokyo');

echo $this->Form->dateTime('published');
```
It will output the following html:
```html
<select name="published[day]">
...
<option value="06" selected="selected">6</option>
...
</select>
<select name="published[hour]">
...
<option value="18" selected="selected">18</option>
...
</select>
```

#### 5. No way to reset default output timezone

By default, `Time::i18Format()` works like the following:
```php
$time = new Time('2016-06-07 03:00:00 +09:00');
$gmtime = clone $time;
$gmtime->setTimezone('UTC');

echo $time;
// 2016/06/07 3:00

echo $gmtime;
// 2016/06/06 18:00
```
 
But after calling `Time::setDefaultOutputTimezone()`, `Time` class no longer behaves like the above, unless I set the protected `_defaultOutputTimezone` property to `null` directly via `ReflectionProperty`.

### Suggestion

I am not sure whether all of the above are intended or not. But at least, it would be difficult to explain this feature at the moment. We need to define what is output timezone, and need to work more for this feature if there are some unintended behaviors.

And it would be better to revert cakephp/app#363 or comment out the related lines by default.

**Edit:** Fixed typo in the example code (4.).

**Comment by ionas** [comment]
This is by design. There is no reasonable base to convert dates with timezones to the same date and another timezone and thus this was omitted, e.g. dates don't really carry enough information about timezones as timezones are about shifting hourly offsets (+09:00 or -03:00).

**Comment by chinpei215** [comment]
How is the SO question related with this issue? The following code displays `2016/06/07` though.
```php
date_default_timezone_set('Asia/Tokyo'); // Change UTC to JST.
Date::setDefaultLocale('ja_JP');
echo new Date();
```

**Comment by ionas** [comment]
> How is the SO question related with this issue? 

The answer is, in my humble opinion: http://stackoverflow.com/a/20289096
As far as I understood: When using date time zones you should use `DateTime` over `Date` even if you only output the date-part ever, and never the hour/minute part.

> The following code displays 2016/06/07 though.

For you this should work without changing the defaults (that maybe should stay to UTC):
```php
echo new Date('now', 'Asia/Tokyo');
```
or with a recent/updated `app.php`...
```php
echo new Date('now', Configure::readOrFail('app.defaultOutputTimezone'));
```
However I'd recommend simply using `DateTime` and not `Date` such as:
```php
Time::setDefaultOutputTimezone('Asia/Tokyo');
echo new Time()->i18nFormat('yyyy-MM-dd');
```

**If all that is no good for**: You can try removing the exception throws and class name checks from `DateFormatTrait` and see if it works for you. ***I had a lot of problems with unit tests conversion of date's timezones*** - because it makes no sense, logically.

**Comment by markstory** [comment]
@chinpei215 Date objects are intended to represent 'calendar dates' and thus don't have a timezone. For example, 2016-06-05 is happens in all timezones eventually even if they don't have it at the same time.

**Comment by chinpei215** [comment]
I updated my original issue, since I noticed another issues. Please reread and rethink it.

**Comment by markstory** [comment]
Effecting the output of `json_encode()` is probably not a desirable side-effect as it can easily cause issues with API based applications.

For 5. we could have a method that removes the output timezone preference. That would at least remove the need to use Reflection to unset the timezone preference.

As for what counts as output, I think the intention and desire I've heard from folks in the community is to make it easier to consistently format dates in a user's preferred timezone. I've mostly heard this related to building HTML interfaces, and less around API responses.

There is a chance that we should consider this to be a `TimeHelper` feature instead of in the Time object itself. I think one reason it was implemented in `Time` was to make use cases like:

```php
<span><?= h($article->created) ?></span>
```

work out of the box.

**Comment by jhli** [comment]
I can change the app To only feature an outcommented line in bootstrap.php.

1. is by design.
2. is debatable but imho by design. Some Angular apps etc will want that. Solution: Use the setter later (in beforeFilter, when its no json response)
3. is bad - makstory, should I prep a PR?
4. is bad as well - again, should I prep a PR?
5. Did you try setting it to UTC? I wanted it to have a setter that can take null originally.

**Comment by ADmad** [comment]
> There is a chance that we should consider this to be a TimeHelper feature instead of in the Time object itself.

This might probably be a better idea. More often than not a user's timezone preference will be fetched from db based on logged in user, so setting default output timezone in bootstrap as suggested in cakephp/app#363 wouldn't be feasible.

Outputs shown in 2 & 3 are something I wouldn't expect either.

**Comment by chinpei215** [comment]
I prefer implementing it in `TimeHelper` too.

But  I understand that it requires more code in templates. So I would like to hear how others say, and even if their conclusion is not the same as mine, I want to comply with it. To be honest, I wouldn't use this feature actually because most japanese people use only JST (GMT+9) in thier applications, and it is very rare to use UTC or other timezones. So it would be better to hear opinions from people who want this feature more frequently. 

@ionas Setting default output timezone to UTC is not the same as setting it to null. Because Time objects have each timezone, and by default, the each timezone is used for each output. But after setting default output timezone to UTC, only UTC is used for output regardless what timezone is set to each time object.

**Comment by ionas** [comment]
I can try to solve:
1. wontfix (use datetime for timezone aware dates)
2. by docs/examples
3. by PR
4. by PR
5. by PR (allow setting to NULL)

**Comment by markstory** [comment]
@ionas I don't think log formatting can be fixed without removing default timezones from `__toString()` as that is the method that is being invoked.

Other than having to write more code what were the drawbacks of doing the default timezone through `TimeHelper`? It would allow the new behavior to be a more opt-in and explicit process.

**Comment by ionas** [comment]
- Loggers could use `->format()` / `->i18nFormat()? calls and specify UTC (or even get the server side default timezone)?

In terms of features: The big drawback aside being more verbose itself is that `TimeHelper` is really limited to view template rendering. 

In terms of *write more code* I can take a look at TimeHelper support - but don't know when my deadlines allow - but I'd keep the current way and just not suggest to disable it by default (app.php/bootstrap.php) and I'd try to patch issues 3. / 4. and 5. myself ASAP.

**Comment by chinpei215** [comment]
@ionas Before starting to patch, you might want to explain your idea. Because, in my opinion, it is very difficult to fix 4.

As you know, we have a two time classes, but only one FormHelper. We need to let FormHelper know which time class should be used. Theoretically, the class should be the same class as which `DateTimeType` class uses.

But now, I can see another annoying issue. Let's assume after submitting the localized form. The posted value would be like the following:
```php
[
    'year' => '2016',
    'month' => '06',
    'day' => '07',
    'hour' => '03',
    'minute' => '00'
]
```
What timezone is used for this array?

You may think that everything goes well if you call `setDefaultOutputTimezone()` in the view layer. But it doesn't. Parsing datetime is done in the controller/model layer. `setDefaultOutputTimezone()` is not called yet. Thus, `DateTimeType` would parse this array as the default timezone, not default output timezone.

**Comment by markstory** [comment]
@ionas I don't think trying to make the loggers aware of time objects is a wise plan. What scenarios did you have in mind where using the helper wouldn't be feasible? One that comes to mind for me, is building flash messages, but I'm not sure how often times are put into those.

**Comment by ionas** [comment]
Being on tight deadlines I had a bit of time to think about this again. There are basically two ways out. The hard way would be fixing issue 2 to 5.

Then I thought about `TimeHelper` again. I think something like this might work:
- don't merge/revert the outputTimezone PR
- add `DateFormatTrait::$_userTimezone` including its getters and setters to save not an output but user timezone. It could also be `i18n::$_userTimezone` instead.
- add `DateFormatTrait::toUserTimezone()` would use above private to convert a time object's timezone.
- add `TimeHelper::setDefaultTimezone($time)` would overwrite the default timezone that is specified via `DateFormatTrait::$_userTimezone` or even `i18n::$_userTimezone`.
- add `TimeHelper::toTimezone($time, $timezone)` to explicitly convert a time object (*OR well formatted string's?*) timezone returning a `Time/cloned FrozenTime` object.

But instead of hacking away I am just suggesting this. What do you like about it? What don't you like? Where do you see ways to do it better/easier?

**Comment by markstory** [comment]
> * don't merge/revert the outputTimezone PR
> * add TimeHelper::setDefaultTimezone($time) would overwrite the default timezone that is specified 

I'm planning on doing these things this weekend. I'm also thinking that TimeHelper could have config options for the default timezone which would make setting the users' preferred timezone simpler as it can be done in the View class/Controller.

**Comment by jhli** [comment]
What do you think about a DateFormatTrait method that reads a global setting and sets timezones by that?

**Comment by jhli** [comment]
nvm, most methods already take that... Maybe no point.

**Comment by markstory** [comment]
Closing as the original changes have been reverted, and a new pull request for the TimeHelper approach has been opened.
